lint.go

```diff
@@ -713,3 +714,81 @@ func CheckEarlyDefer(f *lint.File) {
 	}
 	f.Walk(fn)
 }
+
+func CheckEmptyCriticalSection(f *lint.File) {
+
+	mutexParams := func(s ast.Stmt) (selectorTokens []string, funFullName string, ok bool) {
+		expr, ok := s.(*ast.ExprStmt)
+		if !ok {
+			return nil, "", false
+		}
+		call, ok := expr.X.(*ast.CallExpr)
+		if !ok {
+			return nil, "", false
+		}
+		sel, ok := call.Fun.(*ast.SelectorExpr)
+		if !ok {
+			return nil, "", false
+		}
+
+		// Make sure it's chain of identifiers without any function calls
+		chain := []string{}
+	Loop:
+		for nsel := sel.X; ; {
+			switch s := nsel.(type) {
+			case *ast.Ident:
+				chain = append(chain, s.Name)
+				break Loop
+			case *ast.SelectorExpr:
+				chain = append(chain, s.Sel.Name)
+				nsel = s.X
+			default:
+				return nil, "", false
+			}
+		}
+
+		fn, ok := f.Pkg.TypesInfo.ObjectOf(sel.Sel).(*types.Func)
+		if !ok {
+			return nil, "", false
+		}
+		funName := fn.FullName()
+
+		return chain, funName, true
+	}
+
+	fn := func(node ast.Node) bool {
+		block, ok := node.(*ast.BlockStmt)
+		if !ok {
+			return true
+		}
+		if len(block.List) < 2 {
+			return true
+		}
+		for i := range block.List[:len(block.List)-1] {
+
+			sel1, method1, ok1 := mutexParams(block.List[i])
+			sel2, method2, ok2 := mutexParams(block.List[i+1])
+
+			if !ok1 || !ok2 || len(sel1) != len(sel2) {
+				continue
+			}
+
+			equal := true
+			for i := range sel1 {
+				equal = equal && (sel1[i] == sel2[i])
+			}
+			if !equal {
+				continue
+			}
+
+			if (method1 == "(*sync.Mutex).Lock" && method2 == "(*sync.Mutex).Unlock") ||
```

Comment:
`(interface)` will be fine. It's an interface, the name doesn't matter. The user could also define his own, equivalent interface. Strictly speaking, someone could implement and use the interface without ever using an interface value. We could see a `(*T).Lock()`. However, I think it's better to restrict ourselves to the built-in lock types. It seems more likely that someone uses the Lock/Unlock names but doesn't implement a lock, than that someone is implementing their own mutex.